### PR TITLE
Added Custom Reports API endpoints

### PR DESCRIPTION
**Pull request by rodolfovenancio**

### CODE REVIEW

File: personio-personnel-data-api-oa3.yaml [hunk 1]
```diff
@@ -3983,3 +4105,700 @@ components:
...
+            - point_in_time
+            - historical_data
+            - timeframe
+        employee_status:
+          type: string
+          example: currently_active
+          enum:
+            - currently_active
+            - active_at_given_date
+            - all
```
Comment (raul-mdelfin) [inline personio-personnel-data-api-oa3.yaml:4239]: What does all mean in this context?

File: personio-personnel-data-api-oa3.yaml [hunk 2]
```diff
@@ -3983,3 +4105,700 @@ components:
...
+    SalaryAttribute:
+      type: object
+      properties:
+        attribute_id:
+          type: string
+          example: salary_123
+        data_type:
+          type: string
+          enum:
+            - SALARY
```
Comment (raul-mdelfin) [inline personio-personnel-data-api-oa3.yaml:4614]: Does salary include its type? Fixed/Hourly?

File: personio-personnel-data-api-oa3.yaml [hunk 3]
```diff
@@ -3983,3 +4105,700 @@ components:
...
+      properties:
+        attribute_id:
+          type: string
+          example: ABCXXX
+        data_type:
+          type: string
+          enum:
+            - COMPENSATION
+        amount:
+          type: string
```
Comment (raul-mdelfin) [inline personio-personnel-data-api-oa3.yaml:4636]: Shouldn't the amount be double/float?

File: personio-personnel-data-api-oa3.yaml (hunk 2 above)
Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:4614]: @raul-mdelfin no, the attribute itself is fixed or hourly.

File: personio-personnel-data-api-oa3.yaml (hunk 1 above)
Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:4239]: @raul-mdelfin, these are the returned options values for this field.

File: personio-personnel-data-api-oa3.yaml (hunk 3 above)
Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:4636]: @raul-mdelfin yes, these values are stored as strings.

File: personio-personnel-data-api-oa3.yaml [hunk 4]
```diff
@@ -3983,3 +4105,700 @@ components:
...
+          format: date-time
+          example: "2019-04-10"
+        created_at:
+          type: string
+          format: date
+          example: "2019-04-10"
+        updated_at:
+          type: string
+          format: date
+          example: "2019-04-10"
```
Comment (Macavity) [inline personio-personnel-data-api-oa3.yaml:4262]: For all of these dates, doesn't it make sense to choose a format which contains the Timezone?

Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:4262]: None of these dates have a timezone. We use UTC by default, which is not stored in the DB column.

Comment (Macavity) [reply personio-personnel-data-api-oa3.yaml]: I see - as long as the documentation reflect that, that is just as fine :)

Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml]: I have changed it. Thanks! 🚀

File: personio-personnel-data-api-oa3.yaml [hunk 5]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+          schema:
+            type: array
+            items:
+              type: string
+          required: false
+        - name: status
+          in: query
+          schema:
+            type: string
+          example: "up_to_date"
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2240]: [Nitpick]: I see that you provided examples for some fields and for others you aren't. Could we keep it consistent and provide it for all fields?

File: personio-personnel-data-api-oa3.yaml [hunk 6]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+    get:
+      tags:
+        - Custom Reports
+      parameters:
+        - name: X-Company-Id
+          in: header
+          schema:
+            type: integer
+          required: true
+        - name: report_ids
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2229]: [Question]: If a list of `report_ids` is provided and some of them do not exist, what is the expected response in this case?

File: personio-personnel-data-api-oa3.yaml [hunk 7]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+    get:
+      tags:
+        - Custom Reports
+      parameters:
+        - name: X-Company-Id
+          in: header
+          schema:
+            type: integer
+          required: true
+        - name: reportId
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2268]: [Chore]: Please add a description to all headers & parameters.

File: personio-personnel-data-api-oa3.yaml [hunk 8]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+              schema:
+                $ref: "#/components/schemas/PublicListReportsResponse"
+        500:
+          description: Internal Server Error
+          content:
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/PublicAPIException"
+
+  /company/custom-reports/reports/{reportId}:
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2258]: [Suggestion]: In the previous GET request, the list of report IDs was provided as `report_ids`, here we have `reportId`. Could we maybe have `report_id` here for consistency reasons?

File: personio-personnel-data-api-oa3.yaml (hunk 7 above)
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2268]: [Suggestion]: Same suggestion as above to use `report_id` for consistency.

File: personio-personnel-data-api-oa3.yaml [hunk 9]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+          required: false
+      summary: Get report with items
+      operationId: listReportItems
+      responses:
+        200:
+          description: Ok
+          content:
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/PublicReportResponse"
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2287]: [Question]: What happens if a `report_id` is provided that doesn't exist? Will the response be something like a `400` or `404`? Maybe we could add this case here in the `responses` section.

File: personio-personnel-data-api-oa3.yaml [hunk 10]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/PublicAPIException"
+
+  /company/custom-reports/columns:
+    get:
+      tags:
+        - Custom Reports
+      parameters:
+        - name: X-Company-Id
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2300]: [Chore]: Please add a description to all headers & parameters.

File: personio-personnel-data-api-oa3.yaml [hunk 11]
```diff
@@ -2216,6 +2216,128 @@ paths:
...
+              type: string
+          example: first_name,last_name
+          required: false
+        - name: locale
+          in: query
+          schema:
+            type: string
+          example: "en"
+          required: false
+        - name: report_id
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:2320]: [Suggestion]: Here we are using `report_id` again. Let's make it consistent by using `report_id` everywhere (referring to the previous GET request).

File: personio-personnel-data-api-oa3.yaml [hunk 12]
```diff
@@ -3983,3 +4105,722 @@ components:
         - link
         - tags
         - multiline
+
+    PublicListReportsResponse:
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:4109]: [Question]: This response contains information about pagination. The endpoint definitions you are providing are not describing any `limit` or `offset` fields. Are you going to support pagination for these endpoints? If so, please provide definitions for the `limit` and `offset` fields for all endpoints where you want to allow pagination.

File: personio-personnel-data-api-oa3.yaml [hunk 13]
```diff
@@ -3983,3 +4105,722 @@ components:
...
+        timestamp:
+          type: string
+          format: date-time
+          example: "2020-01-01T00:00:00Z"
+        errors:
+          type: array
+          items:
+            $ref: "#/components/schemas/ErrorDetails"
+
+    ErrorDetails:
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:4806]: [Chore]: Please provide example values for all defined fields of this component.

File: personio-personnel-data-api-oa3.yaml [hunk 14]
```diff
@@ -3983,3 +4105,722 @@ components:
...
+        details:
+          type: string
+        type:
+          type: string
+        _meta:
+          type: array
+          items:
+            $ref: "#/components/schemas/ErrorMeta"
+
+    ErrorMeta:
```
Comment (dinomite94) [inline personio-personnel-data-api-oa3.yaml:4820]: [Chore]: Please provide example values for all defined fields of this component.

**Review by dinomite94 — CHANGES_REQUESTED** [review]

File: personio-personnel-data-api-oa3.yaml (hunk 9 above)
Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:2287]: Good catch, yes, that returns a 404 👍

File: personio-personnel-data-api-oa3.yaml (hunk 12 above)
Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:4109]: We do not support pagination now. That will be added in the next version, but these fields are being used for the client to know the number of results of the request.

File: personio-personnel-data-api-oa3.yaml (hunk 6 above)
Comment (rodolfovenancio) [reply personio-personnel-data-api-oa3.yaml:2229]: No exception is expected here. If the reports do not exist, they are not being returned.

**Review by dinomite94 — APPROVED** [review]

**Pull request merged**